File: src/io/read_raw_input.nim
import posix
import posix/termios as posix_termios

proc cfMakeRaw(termios: ptr Termios) {.importc: "cfmakeraw",
    header: "<termios.h>".}

proc readRawInput*(): string =
  var oldTerm, newTerm: Termios
  discard tcGetAttr(STDIN_FILENO.cint, addr oldTerm)
  newTerm = oldTerm
  cfMakeRaw(addr newTerm)
  discard tcSetAttr(STDIN_FILENO.cint, TCSANOW, addr newTerm)

  var buf: array[2, char]
  let bytesRead = read(STDIN_FILENO.cint, addr buf[0], 1)

  discard tcSetAttr(STDIN_FILENO.cint, TCSANOW, addr oldTerm)

  if bytesRead > 0:
    result = $buf[0]
  else:
    result = ""

File: src/io/read_raw_input.rs
use std::io::Read;
use nix::sys::termios;
use std::os::fd::BorrowedFd;

pub fn read_raw_input() -> Option<String> {
    let stdin_fd = std::io::stdin();

    // Save old terminal settings
    let fd = unsafe { BorrowedFd::borrow_raw(0) };
    let old_term = termios::tcgetattr(fd).ok()?;

    // Set raw mode
    let mut new_term = old_term.clone();
    termios::cfmakeraw(&mut new_term);
    termios::tcsetattr(fd, termios::SetArg::TCSANOW, &new_term).ok()?;

    // Read a single byte
    let mut buf = [0u8; 1];
    let result = stdin_fd.lock().read(&mut buf);

    // Restore old terminal settings
    termios::tcsetattr(fd, termios::SetArg::TCSANOW, &old_term).ok()?;

    match result {
        Ok(n) if n > 0 => Some(String::from_utf8_lossy(&buf[..n]).to_string()),
        _ => None,
    }
}
